### PR TITLE
docs(hmr): document some HMR related functions

### PR DESCRIPTION
**Pull request by alicewriteswrongs**
This adds JSDoc and inline comments to some HMR related code, mainly just recording my understanding of what this code does / is supposed to do.

## What is the current behavior?

Most of this stuff isn't documented at all!

## What is the new behavior?

Now it is 😅

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Testing

This should be a comment-only change so I don't think any functionality should be meaningfully impacted.

### CODE REVIEW

**Review by alicewriteswrongs — COMMENTED** [review]
just two notes

File: src/dev-server/client/hmr-components.ts [hunk 1]
```diff
@@ -1,36 +1,64 @@
...
       updatedTags.push(cmpTagName);
     }
   }
 
-  if (hasShadowRoot(elm)) {
-    hmrComponent(updatedTags, elm.shadowRoot as any, versionId, cmpTagName);
+  // recur into any elements which have a shadow root
+  if (hasShadowRoot(element)) {
+    hmrComponent(updatedTags, element.shadowRoot as any, versionId, cmpTagName);
   }
```
[the remark this answers is not in the document]
Comment (rwaskiewicz) [reply src/dev-server/client/hmr-components.ts:57]: Can't walk the tree if you're not allowed in 😅

**Review by rwaskiewicz — APPROVED** [review]

File: src/dev-server/client/hmr-components.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (alicewriteswrongs) [reply src/dev-server/client/hmr-components.ts:57]: haha true enough!

**Review by tanner-reits — APPROVED** [review]
Comments look good. So much `any` casting in these files 😭

**Pull request merged**